### DISCUSSION TITLE
Suppor for relative path using QUERY_BASE_PATH

### DISCUSSION THREAD

**Discussion by Sidhartha-Biswal** [jaegertracing/jaeger#5042] [Q&A]
Hi,

Does QUERY_BASE_PATH environment variable support relative-path case while using reverse proxy? Can it be set using QUERY_STATIC_FILES?

What are the possible values for QUERY_STATIC_FILES environment variable?

**Comment by yurishkuro** [comment]
Static files flag is a legacy remnant from the time when ui assets were packaged separately from the binary and you could point the binary to different locations for those. Since the assets are bundled inside the binary now this flag is not needed.

**Reply by Sidhartha-Biswal** [reply]
Hi @yurishkuro,

How do we configure the QUERY_BASE_PATH so that the relative path will be used for loading the static files like css and js when reverse proxy is used?

**Reply by yurishkuro** [reply]
I don't understand what you're asking. Static files have nothing to do with QUERY_BASE_PATH. 

QUERY_BASE_PATH simply makes the query service respond to a different URL, e.g. instead of the default `/api/...` you could run it with `export QUERY_BASE_PATH=/blah` and the query service will respond to `/blah/api/...`

**Reply by Sidhartha-Biswal** [reply]
Hi @yurishkuro,

The js/css/.. resources are taking the absolute path and the UI shows error when using a reverse proxy with a prefix. The files related to the loading of the UI, for example: CSS and JS files are giving 404 errors.

 Is it possible to make the download of js/css/.. resources with relative paths while using reverse proxy? could it be configurable?

**Reply by yurishkuro** [reply]
```
$ go run -tags=ui ./cmd/all-in-one --query.base-path=/somepath
```
In this example Jaeger is started with a non-default base path. There are no issues with web assets. You may have an issue with your reverse proxy configuration.

**Reply by Sidhartha-Biswal** [reply]
Hi @yurishkuro,
The issue is coming up only when a rewrite is done for the QUERY_BASE_PATH.
So is it possible to rewrite the QUERY_BASE_PATH or not?

**Reply by yurishkuro** [reply]
please provide reproducible example

**Reply by Sidhartha-Biswal** [reply]
Hi @yurishkuro,
Suppose QUERY_BASE_PATH is "abc/xyz" and we are giving "demo/viewer/jaeger/v1/abc/xyz" as the rewrite value then the UI is not coming up and it shows loading error for the index.js and index.css.
Is there any limitation with respect to the character length for the QUERY_BASE_PATH value?

**Reply by yurishkuro** [reply]
this is not a reproducible example. If you're running an open source proxy then provide full relevant configuration.

There is no length limit.

**Reply by kierangillblueberry** [reply]
@yurishkuro We have a similar requirement. We proxy all of our local dev traffic behind a URL (e.g., `local.example.dev`). 

```
# Caddyfile
local.example.dev {
    reverse_proxy website:8000
    tls /etc/caddy/bbp.crt /etc/caddy/bbp.key
}

local.example.dev/other-service* {
    reverse_proxy service:12345
    tls /etc/caddy/bbp.crt /etc/caddy/bbp.key
}
```

We'd like to proxy all traffic to `local.example.dev/traces` to `jaeger`.

```
local.example.dev/traces* {
    reverse_proxy traces:16686
    tls /etc/caddy/bbp.crt /etc/caddy/bbp.key
}
```

Is this supported with the Docker image `jaegertracing/all-in-one`? E.g.:

```
  traces:
    image: jaegertracing/all-in-one:1.32
    ports:
      - 16686:16686
      - 6831:6831
    environment:
      BASE_URL: /traces
```

Without being able to set the base url, navigating to `local.example.dev` results in the error:

```
Refused to apply style from 'https://local.example.dev/static/css/main.c3c64efc.chunk.css' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled
```

`/static` won't resolve to `jaeger`, it instead resolves to our `website` service.

**Reply by yurishkuro** [reply]
Works in this example: https://github.com/jaegertracing/jaeger/tree/main/examples/reverse-proxy

**Reply by Sidhartha-Biswal** [reply]
Hi @yurishkuro,

The traffic flow is gateway>virtual service>gateway>jaeger query.

We are setting the query_base_path as "/jaeger/viewer" and then doing the rewrite using the virtual service configuration mentioned below. 

- match:
    - uri:
        prefix: /jaeger/viewer
    rewrite:
       uri: /hub/jaeger-query/v1/jaeger/viewer
    retries:
      attempts: 3
      perTryTimeout: 15s
      retryOn: connect-failure,refused-stream,reset,unavailable,cancelled,resource-exhausted
    route:
    - destination:
        host: api-gateway
        port:
          number: 8080
    timeout: 15s
    
Please find the attached screenshot. The url is http://hostname/hub/jaeger-query/v1/jaeger/viewer
    
![jaegeruierror](https://github.com/jaegertracing/jaeger/assets/119291884/af1cdaf5-4e59-495a-b007-54fc997f7e2e)

We are getting issue in the Jaeger UI.The index.html is getting loaded but the static files like index.js, index.css are showing loading error(404 not found) in the console tab of the browser.

If we click on the "Back Home" button then the url is changing as http://hostname/jaeger/viewer and the UI is coming up properly.

So we would like to understand if Jaeger supports the rewrite of the query_base_path using virtual service?

Please provide your opinion.

**Reply by yurishkuro** [reply]
It didn't work for me either when public and internal prefix was different, why don't you make it the same?

**Reply by Sidhartha-Biswal** [reply]
Hi @yurishkuro,

We want public and internal prefix to be different. Does Jaeger support that?

We set the query_base_path as "/hub/jaeger-query/v1/jaeger/viewer" and then modified the virtual service configuration as mentioned below. But it did not work. Please suggest your opinion about what could be the issue.

match:
    uri:
    prefix: /hub/jaeger-query/v1/jaeger/viewer
    rewrite:
    uri: /hub/jaeger-query/v1/jaeger/viewer
    retries:
    attempts: 3
    perTryTimeout: 15s
    retryOn: connect-failure,refused-stream,reset,unavailable,cancelled,resource-exhausted
    route:
    destination:
    host: api-gateway
    port:
    number: 8080
    timeout: 15s

**Reply by yurishkuro** [reply]
Different public/private prefixes do not work today (it would require configuring them separately in jaeger-query, which is not supported). Theoretically you can make it work by doing rewriting, but it's pretty tricky to do for an AJAX app which constructs URLs on the fly from Javascript.

**Reply by Sidhartha-Biswal** [reply]
Hi @yurishkuro,
We would like to know if this is a bug in Jaeger and will this use case be supported if any modifications are done in the Jaeger source code or if it is an issue with the client side configuration/code. Please share your opinion.

**Reply by yurishkuro** [reply]
Bug is subjective - I don't see it as a bug, Jaeger is doing what it promises to do, and it has no control over layers you put on top of it. Like I said, I think it's possible to support your use case by introducing two independent "prefixes", local (current implementation) and external (to be used from the UI). But not supporting it is not a bug, just a missing feature.

You may open a ticket to describe your use case and why it's important (and cannot be done using the existing support), and if there is interest in the community it may be implemented (I think changes are relatively simple).
